File: app/src/main/java/com/example/advancedprayertimes/Logic/Entities/PrayerTimePackageAbstractClass.java
package com.example.advancedprayertimes.Logic.Entities;

import com.example.advancedprayertimes.Logic.Enums.EPrayerTimeMomentType;
import com.example.advancedprayertimes.Logic.Enums.EPrayerTimeType;

import java.time.LocalDateTime;

public abstract class PrayerTimePackageAbstractClass
{
    public abstract String getDate();

    public abstract void setDate (String date);

    public abstract LocalDateTime getFajrTime();
    public abstract void setFajrTime(LocalDateTime fajrTime);

    public abstract LocalDateTime getSunriseTime();
    public abstract void setSunriseTime(LocalDateTime sunriseTime);

    public abstract LocalDateTime getDuhaTime();
    public abstract void setDuhaTime(LocalDateTime duhaTime);

    public abstract LocalDateTime getDhuhrTime();
    public abstract void setDhuhrTime(LocalDateTime dhuhrTime);

    public abstract LocalDateTime getAsrTime();
    public abstract void setAsrTime(LocalDateTime asrTime);

    public abstract LocalDateTime getMithlaynTime();
    public abstract void setMithlaynTime(LocalDateTime mithlaynTime);

    public abstract LocalDateTime getAsrKarahaTime();

    public abstract void setAsrKarahaTime(LocalDateTime asrKarahaTime);

    public abstract  LocalDateTime getMaghribTime();
    public abstract void setMaghribTime(LocalDateTime maghribTime);

    public abstract LocalDateTime getIshtibaqAnNujumTime();
    public abstract void setIshtibaqAnNujumTime(LocalDateTime ishtibaqAnNujum);

    public abstract LocalDateTime getIshaTime();
    public abstract void setIshaTime(LocalDateTime ishaTime);

    public LocalDateTime GetTimeByType(EPrayerTimeType prayerType, EPrayerTimeMomentType prayerTypeTimeType)
    {
        if((prayerType == EPrayerTimeType.Isha && prayerTypeTimeType == EPrayerTimeMomentType.End)
                || (prayerType == EPrayerTimeType.Fajr && prayerTypeTimeType == EPrayerTimeMomentType.Beginning))
        {
            return this.getFajrTime();
        }
        else if(prayerType == EPrayerTimeType.Fajr && prayerTypeTimeType == EPrayerTimeMomentType.End)
        {
            return this.getSunriseTime();
        }
        else if(prayerType == EPrayerTimeType.Duha && prayerTypeTimeType == EPrayerTimeMomentType.Beginning)
        {
            return this.getDuhaTime();
        }
        else if(prayerType == EPrayerTimeType.Duha && prayerTypeTimeType == EPrayerTimeMomentType.End)
        {
            LocalDateTime returnLocalDateTime = this.getDhuhrTime();
            returnLocalDateTime = returnLocalDateTime.minusHours(1);

            return returnLocalDateTime;
        }
        else if(prayerType == EPrayerTimeType.Dhuhr && prayerTypeTimeType == EPrayerTimeMomentType.Beginning)
        {
            return this.getDhuhrTime();
        }
        else if((prayerType == EPrayerTimeType.Dhuhr && prayerTypeTimeType == EPrayerTimeMomentType.End)
                || (prayerType == EPrayerTimeType.Asr && prayerTypeTimeType == EPrayerTimeMomentType.Beginning))
        {
            return this.getAsrTime();
        }
        else if((prayerType == EPrayerTimeType.Asr && prayerTypeTimeType == EPrayerTimeMomentType.End)
                || (prayerType == EPrayerTimeType.Maghrib && prayerTypeTimeType == EPrayerTimeMomentType.Beginning))
        {
            return this.getMaghribTime();
        }
        else if((prayerType == EPrayerTimeType.Maghrib && prayerTypeTimeType == EPrayerTimeMomentType.End)
                || (prayerType == EPrayerTimeType.Isha && prayerTypeTimeType == EPrayerTimeMomentType.Beginning))
        {
            return this.getIshaTime();
        }

        return null;
    }
}

File: app/src/main/kotlin/com/example/advancedprayertimes/logic/api_entities/PrayerTimePackageAbstractClass.kt
package com.example.advancedprayertimes.logic.api_entities

import com.example.advancedprayertimes.logic.enums.EPrayerTimeType
import com.example.advancedprayertimes.logic.enums.EPrayerTimeMomentType
import java.time.LocalDate
import java.time.LocalTime
import java.util.AbstractMap

abstract class PrayerTimePackageAbstractClass
{
    abstract var date: LocalDate?
    abstract var fajrTime: LocalTime?
    abstract var sunriseTime: LocalTime?
    abstract var duhaTime: LocalTime?
    abstract var dhuhrTime: LocalTime?
    abstract var asrTime: LocalTime?
    abstract var mithlaynTime: LocalTime?
    abstract var asrKarahaTime: LocalTime?
    abstract var maghribTime: LocalTime?
    abstract var ishtibaqAnNujumTime: LocalTime?
    abstract var ishaTime: LocalTime?

    fun getTimeByType(prayerType: EPrayerTimeType, prayerTypeTimeType: EPrayerTimeMomentType): LocalTime? {

        return when(AbstractMap.SimpleEntry(prayerType, prayerTypeTimeType))
        {
            AbstractMap.SimpleEntry(EPrayerTimeType.Isha, EPrayerTimeMomentType.End),
            AbstractMap.SimpleEntry(EPrayerTimeType.Fajr, EPrayerTimeMomentType.Beginning) -> this.fajrTime

            AbstractMap.SimpleEntry(EPrayerTimeType.Fajr, EPrayerTimeMomentType.End) -> this.sunriseTime

            AbstractMap.SimpleEntry(EPrayerTimeType.Duha, EPrayerTimeMomentType.Beginning) -> this.duhaTime

            AbstractMap.SimpleEntry(EPrayerTimeType.Duha, EPrayerTimeMomentType.End) -> this.dhuhrTime?.minusHours(1)

            AbstractMap.SimpleEntry(EPrayerTimeType.Dhuhr, EPrayerTimeMomentType.Beginning) -> this.dhuhrTime

            AbstractMap.SimpleEntry(EPrayerTimeType.Dhuhr, EPrayerTimeMomentType.End),
            AbstractMap.SimpleEntry(EPrayerTimeType.Asr, EPrayerTimeMomentType.Beginning) -> this.asrTime

            AbstractMap.SimpleEntry(EPrayerTimeType.Asr, EPrayerTimeMomentType.End),
            AbstractMap.SimpleEntry(EPrayerTimeType.Maghrib, EPrayerTimeMomentType.Beginning) -> this.maghribTime

            AbstractMap.SimpleEntry(EPrayerTimeType.Maghrib, EPrayerTimeMomentType.End),
            AbstractMap.SimpleEntry(EPrayerTimeType.Isha, EPrayerTimeMomentType.Beginning) -> this.ishaTime

            else -> null
        }
    }
}